File: GitToolBox/src/main/java/zielu/gittoolbox/blame/BlameSubscriber.java
package zielu.gittoolbox.blame;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.messages.MessageBusConnection;
import git4idea.repo.GitRepository;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.cache.PerRepoInfoCache;
import zielu.gittoolbox.cache.PerRepoStatusCacheListener;
import zielu.gittoolbox.cache.RepoInfo;

class BlameSubscriber {

  BlameSubscriber(@NotNull Project project) {
    MessageBusConnection connection = project.getMessageBus().connect(project);
    connection.subscribe(PerRepoInfoCache.CACHE_CHANGE, new PerRepoStatusCacheListener() {
      @Override
      public void stateChanged(@NotNull RepoInfo info, @NotNull GitRepository repository) {
        BlameCache.getExistingInstance(project).ifPresent(cache -> cache.refreshForRoot(repository.getRoot()));
      }
    });
    connection.subscribe(BlameCache.CACHE_UPDATES, new BlameCacheListener() {
      @Override
      public void cacheUpdated(@NotNull VirtualFile file, @NotNull BlameAnnotation annotation) {
        BlameService.getExistingInstance(project).ifPresent(service -> service.blameUpdated(file, annotation));
      }

      @Override
      public void invalidated(@NotNull VirtualFile file) {
        BlameService.getExistingInstance(project).ifPresent(service -> service.invalidate(file));
      }
    });
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/blame/BlameSubscriber.kt
package zielu.gittoolbox.blame

import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import git4idea.repo.GitRepository
import zielu.gittoolbox.cache.PerRepoInfoCache
import zielu.gittoolbox.cache.PerRepoStatusCacheListener
import zielu.gittoolbox.cache.RepoInfo

internal class BlameSubscriber(project: Project) {
  init {
    val connection = project.messageBus.connect(project)
    connection.subscribe(PerRepoInfoCache.CACHE_CHANGE, object : PerRepoStatusCacheListener {
      override fun stateChanged(info: RepoInfo, repository: GitRepository) {
        BlameCache.getExistingInstance(project).ifPresent {
          cache: BlameCache -> cache.refreshForRoot(repository.root)
        }
      }
    })
    connection.subscribe(BlameCache.CACHE_UPDATES, object : BlameCacheListener {
      override fun cacheUpdated(file: VirtualFile, annotation: BlameAnnotation) {
        BlameService.getExistingInstance(project).ifPresent {
          service: BlameService -> service.blameUpdated(file, annotation)
        }
      }

      override fun invalidated(file: VirtualFile) {
        BlameService.getExistingInstance(project).ifPresent {
          service: BlameService -> service.invalidate(file)
        }
      }
    })
  }
}
